framework/wazuh/core/batcher/batcher.py

```diff
@@ -0,0 +1,196 @@
+
+from wazuh.core.batcher.buffer import Buffer
+from wazuh.core.batcher.timer import TimerManager
+from wazuh.core.batcher.mux_demux import MuxDemuxQueue, Message
+from wazuh.core.batcher.config import BatcherConfig
+
+logger = logging.getLogger('wazuh-comms-api')
+
+WAIT_TIME_BETWEEN_QUEUE_READ = 0.01
+
+
+class Batcher:
+    """
+    Manages the batching of messages from a MuxDemuxQueue and sends them in bulk to an indexer.
+
+    Parameters
+    ----------
+    queue : MuxDemuxQueue
+        The MuxDemuxQueue instance for message multiplexing and demultiplexing.
+    config : BatcherConfig
+        Configuration parameters for batching, such as maximum elements and size.
+    """
+    def __init__(self, queue: MuxDemuxQueue, config: BatcherConfig):
+        self.q: MuxDemuxQueue = queue
+        self._buffer: Buffer = Buffer(max_elements=config.max_elements, max_size=config.max_size)
+        self._timer: TimerManager = TimerManager(max_time_seconds=config.max_time_seconds)
+        self._shutdown_event: Optional[asyncio.Event] = None
+
+    async def _get_from_queue(self) -> Message:
+        """
+        Retrieves a message from the mux queue. If the queue is empty, waits and retries until a message is received
+        or the task is cancelled.
+
+        Returns
+        -------
+        Message
+            The message retrieved from the mux queue.
+
+        Raises
+        ------
+        asyncio.CancelledError
+            If the task is cancelled during the operation.
+        """
+        message = None
+        while True:
+            try:
+                message = self.q.receive_from_mux(block=False)
+                return message
+            except queue.Empty:
+                await asyncio.sleep(WAIT_TIME_BETWEEN_QUEUE_READ)
+            except asyncio.CancelledError:
+                if message is not None:
+                    self.q.send_to_mux(uid=message.uid, msg=message.msg)
+                break
+
+    async def _send_buffer(self, events: List[Message]):
+        """
+        Sends a batch of messages to the indexer in bulk. Updates the demux queue with the response messages.
+
+        Parameters
+        ----------
+        events : List[Message]
+            The list of messages to be sent in bulk.
+
+        Raises
+        ------
+        Exception
+            If an error occurs during the sending of the buffer.
+        """
+        try:
+            async with get_indexer_client() as indexer_client:
+                list_of_uid: List[uuid.UUID] = [event.uid for event in events]
+                bulk_list: List[BulkDoc] = [
+                    BulkDoc.create(index=indexer_client.events.INDEX, doc_id=None, doc=event.msg) for event in events
+                ]
+
+                response = await indexer_client.events.bulk(data=bulk_list)
+
+                for response_item, uid in zip(response["items"], list_of_uid):
+                    response_item_msg = response_item["create"]
```

Comment:
This shouldn't be hardcoded, because the bulk operation could include `update` statements. Otherwise we are limiting the class to indexing data only, and it may be useful in other scenarios.

Reply:
Changed in https://github.com/wazuh/wazuh/commit/7344dcea68aa1f5d451757e21ea6e02cc5e61448.